server/src/main/java/org/elasticsearch/rest/action/search/HttpChannelTaskHandler.java

```diff
@@ -0,0 +1,158 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.rest.action.search;
+
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.action.ActionRequest;
+import org.elasticsearch.action.ActionResponse;
+import org.elasticsearch.action.ActionType;
+import org.elasticsearch.action.admin.cluster.node.tasks.cancel.CancelTasksRequest;
+import org.elasticsearch.client.Client;
+import org.elasticsearch.client.node.NodeClient;
+import org.elasticsearch.http.HttpChannel;
+import org.elasticsearch.tasks.Task;
+import org.elasticsearch.tasks.TaskId;
+import org.elasticsearch.tasks.TaskListener;
+
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.CopyOnWriteArraySet;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicReference;
+
+/**
+ * This class keeps track of which tasks came in from which {@link HttpChannel}, by allowing to associate
+ * an {@link HttpChannel} with a {@link TaskId}, and also removing the link once the task is complete.
+ * Additionally, it accepts a consumer that gets called whenever an http channel gets closed, which
+ * can be used to cancel the associated task when the underlying connection gets closed.
+ */
+final class HttpChannelTaskHandler {
+    final Map<HttpChannel, CloseListener> httpChannels = new ConcurrentHashMap<>();
+
+    <Response extends ActionResponse> void execute(NodeClient client, HttpChannel httpChannel, ActionRequest request,
+                                                   ActionType<Response> actionType, ActionListener<Response> listener) {
+        //0: initial state, 1: either linked or already unlinked without being linked first, 2: first linked and then unlinked
+        //link can only be done if it's the first thing that happens. unlink will only happen if link was done first.
+        AtomicInteger link = new AtomicInteger(0);
```

Comment:
I only see two states used (0 and >0) so maybe an AtomicBoolean is enough ?

Reply:
been there, tried that, multiple times, because I had the same thought. The problem is that we only want link when unlink has not been performed yet, and unlink only when link has already been performed (optional). This is the key statement: `if (link.getAndIncrement() > 0)` which cannot be achieved with a `compareAndSet` using `AtomicBoolean`, as we only do the unlink if link has already happened, but in any case we change the state. We could probably achieve a similar behaviour with atomic boolean if we don't use compare and set, which would be ok as we are in a synchronized block, but I am not sure that we would be able to still skip unlink when link is not yet performed.

0: link has not been called, unlink has not been called. link is allowed, unlink is not performed as we know the task can't be found
1: link has been called, or unlink has been called: if link is called, it is not performed as unlink has already been called. if unlink is called, it is performed as link has been called. This works under the assumption that link / unlink are only called one time each.
2: link has been called, unlink has been called